webdav: allow specification of a different 'geturl' than upload url

fir4 opened the pull request:
Tentative fix for #7914 

Previously the WebDAV Image Uploader would use the provided URL both for upload, and external reference. Unfortunately, this just doesn't work for some systems such as Owncloud, where the WebDAV path is relative to the user: it would require to be logged in as the same user than Grafana to view the image.

This PR adds an additional "geturl" parameter that allows to specify a different base URL than the upload URL. It has no effect when the geturl parameter is not defined.

Example Owncloud 9.1 config (actual URL may vary depending on version) : 
 * Create an user and folder for Grafana in Owncloud
 * Enable public link sharing on the folder and note its ID in URL
 * Point Grafana WebDAV URL at owncloud/remote.php/webdav/FolderName, and configure user/pass accoridngly
 * Point Grafana WebDAV GetURL at owncloud/index.php/apps/files_sharing/ajax/publicpreview.php?x=9999&y=9999&scalingup=0&a=true&t=[FOLDERID]&file=

This is the first time I remotely touch a go program, so even though it's a simple modification, please review it carefully !

torkelo commented:
maybe change option to public_url ? and add it to defaults.ini and sample.ini and docs, and make it optional (so no existing setups are affected by this change)

fir4 committed:
webdav: allow specification of a different public_url than upload url

fir4 commented:
Sorry, i messed up with my PR fork and GitHub autoclosed the PR.
Changed "geturl" to "public_url", added doc and sample lines in sample/defaults.ini
Also ran gofmt so that CircleCI won't fail this time (hopefully !). I retested with and without the public_url option to make sure everything was OK.

torkelo commented on pkg/components/imguploader/webdavuploader.go:
```diff
@@ -53,13 +55,18 @@ func (u *WebdavUploader) Upload(pa string) (string, error) {
 
-	return url.String(), nil
+	if u.public_url != "" {
+		return (u.public_url + filename), nil
```
this should use proper url concatenation to account for trailing or no trialing slash

fir4 replied:
Well i planned initially to use path concatenation, but this always adds a /. This doesn't work out for GET arguments. Do you have any suggestion ?

torkelo commented on docs/sources/installation/configuration.md:
```diff
@@ -635,6 +635,9 @@ Secret key. e.g. AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
 Url to where Grafana will send PUT request with images
 
+### public_url
+Url to send to users in notifications, directly appended with the resulting uploaded file name
```
would be good to mention that this setting is optional, and by default you only need to set url

torkelo replied:
ping

daniellee commented:
Merged manually with a few small changes. Thank you @fir4 for taking the time to add this feature!
